Add .NET 9 target

mus65 opened the pull request:
I had to disable a SonarSource analyzer and made use of the new `Lock` type, see separate Commits.

CI is expected to fail because of missing .NET 9 binaries in AppVeyor. Build and tests pass locally.

Can be merged when .NET 9 final is released and AppVeyor supports it.

mus65 committed:
Disable SonarSource S3236

This following change in the runtime now causes this analyzer
to complain about some Debug.Assert calls which doesn't make sense.

https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#debugassert-now-reports-assert-condition-by-default
https://rules.sonarsource.com/csharp/RSPEC-3236/

mus65 committed:
Add .NET 9 target

mus65 committed:
make use of .NET 9 Lock type

see https://github.com/dotnet/runtime/issues/34812

Rob-Hague commented:
Can we have a basic Lock polyfill to save on the #ifdefs? e.g.

```c#
#if !NET9_0_OR_GREATER

namespace Renci.SshNet.Common;

internal sealed class Lock
{
    public bool TryEnter() => Monitor.TryEnter(this);
    public void Exit() => Monitor.Exit(this);
}

```

mus65 committed:
Define own Lock type to avoid ifdefs

mus65 commented:
Makes sense, pushed.

mus65 committed:
Merge remote-tracking branch 'upstream/develop' into net9

mus65 committed:
revert irrelevant style changes

drieseng commented:
@mus65, you'll also want to update **global.json**.

drieseng commented on .editorconfig:
```diff
@@ -183,6 +183,9 @@ dotnet_diagnostic.S2971.severity = none
 dotnet_diagnostic.S3218.severity = none
 
+# S3236: Remove this argument from the method call; it hides the caller information.
+dotnet_diagnostic.S3236.severity = none
```
Where was this (falsely) reported?

mus65 replied:
I mentioned this in the commit (should have written it in the PR in hindsight). This causes false positives with `Debug.Assert` when a custom message is supplied. This is only a problem since .NET 9 because of this change: https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#debugassert-now-reports-assert-condition-by-default

drieseng commented on appveyor.yml:
```diff
@@ -19,15 +19,15 @@ for:
     - echo build
-    - dotnet build -f net8.0 -c Debug test/Renci.SshNet.Tests/Renci.SshNet.Tests.csproj
-    - dotnet build -f net8.0 -c Debug test/Renci.SshNet.IntegrationTests/Renci.SshNet.IntegrationTests.csproj
+    - dotnet build -f net9.0 -c Debug test/Renci.SshNet.Tests/Renci.SshNet.Tests.csproj
```
Do we want to switch from .NET 8.0 (LTS) to .NET 9.0 (STS, currently still in preview)?
Does AppVeyor even support .NET 9.0?

mus65 replied:
> Do we want to switch from .NET 8.0 (LTS) to .NET 9.0 (STS, currently still in preview)?

Imho it makes sense to test against the latest .NET to catch any regressions. Would even be better to test against all of them, but that's probably not worth the CI overhead. That's up to you.

I didn't want this to be merged while .NET 9 is still in preview, that's why it's still draft.

> Does AppVeyor even support .NET 9.0?

it doesn't. I'm just hoping that it supports it when .NET 9 final is released. If not, the SDK could probably be installed manually somehow.

drieseng commented on test/Renci.SshNet.AotCompatibilityTestApp/Renci.SshNet.AotCompatibilityTestApp.csproj:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net8.0</TargetFramework>
+    <TargetFramework>net9.0</TargetFramework>
```
Do we want this?

Rob-Hague replied:
Fine either way

drieseng commented on test/Renci.SshNet.Benchmarks/Renci.SshNet.Benchmarks.csproj:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net8.0</TargetFramework>
+    <TargetFramework>net9.0</TargetFramework>
```
Do we want this?

Rob-Hague replied:
Yes

drieseng commented on test/Renci.SshNet.IntegrationBenchmarks/Renci.SshNet.IntegrationBenchmarks.csproj:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net8.0</TargetFramework>
+    <TargetFramework>net9.0</TargetFramework>
```
Do we want this?

Rob-Hague replied:
Yes

drieseng commented on test/Renci.SshNet.IntegrationTests/Renci.SshNet.IntegrationTests.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>net48;net8.0</TargetFrameworks>
+    <TargetFrameworks>net48;net9.0</TargetFrameworks>
```
Do we want this?

Rob-Hague replied:
Let's keep a net8.0 target just for local testing

```suggestion
    <TargetFrameworks>net48;net8.0;net9.0</TargetFrameworks>
```

drieseng commented:
@Rob-Hague, are planning on merging this before the release of .NET 9.0? In theory, there may even be API changes - I'm thinking of the **Lock** class - between now and the release. It's highly unlikely though.

mus65 committed:
Merge remote-tracking branch 'upstream/develop' into net9

mus65 committed:
update global.json

mus65 commented:
@drieseng @Rob-Hague I updated global.json . I had to reference RC1 explicitly though, so I have to change it when final is released.

I left it a draft because it's still preview and wouldn't expect it to be merged before .NET 9 final. RC1 is already released, so I wouldn't expect any more API changes.

Rob-Hague commented:
Yeah, we will merge once .NET 9 is released

Newer targets can be used on appveyor with apt-get, as in the opposite of #1352

mus65 committed:
Keep net8.0 target in IntegrationTests

Co-authored-by: Rob Hague <rob.hague00@gmail.com>

mus65 committed:
Merge remote-tracking branch 'upstream/develop' into net9

mus65 committed:
fix Package Downgrade Warning

for some reason this happens starting with .NET 9.0 RC2:

/home/mus/git/SSH.NET/test/Renci.SshNet.IntegrationTests/Renci.SshNet.IntegrationTests.csproj : error NU1605:
      Warning As Error: Detected package downgrade: BouncyCastle.Cryptography from 2.4.0 to 2.3.1. Reference the package directly from the project to select a different version.
       Renci.SshNet.IntegrationTests -> SSH.NET 1.0.0 -> BouncyCastle.Cryptography (>= 2.4.0)
       Renci.SshNet.IntegrationTests -> Testcontainers 3.10.0 -> BouncyCastle.Cryptography (>= 2.3.1)

mus65 committed:
update global.json to RC2

mus65 committed:
Merge remote-tracking branch 'upstream/develop' into net9

mus65 committed:
update global.json to .NET 9 GA

mus65 commented:
This is now only blocked by AppVeyor not supporting .NET 9. apt-get is not going work either because the dotnet9 package is not available in Ubuntu 22.04 (see https://launchpad.net/ubuntu/+source/dotnet9 ).

@Rob-Hague has there been any thought about moving to GitHub actions?

Rob-Hague commented:
I seem to remember @WojciechNagorski trying it but I can't find a PR so may be imagining

mus65 committed:
Merge remote-tracking branch 'upstream/develop' into net9

Rob-Hague commented:
Feel free to give a go though

mus65 committed:
update GitHub Actions for .NET 9

mus65 committed:
Merge remote-tracking branch 'upstream/develop' into net9
